### PR TITLE
Translations support V2

### PR DESCRIPTION
**Pull request by math07**
Hey Shinnok,
On startup, Qt detects your system language and Johnny is translated to it if the translation is available. I also made the french translation and refactored a function with defines that didn't work with translation and it is now easier to read.

However, I will commit the option to change language in the settings when I fix the listWidget layout problem. You can wait until this is done before merging my pull request.

### CODE REVIEW

**Comment by math07** [comment]
New update about this pull request on the list:
http://openwall.com/lists/john-dev/2015/04/19/5
and
http://openwall.com/lists/john-dev/2015/04/19/10

**Comment by math07** [comment]
I added dynamic language switching in my latest commit. A new class Translator is now responsible of handling translation. My goal was to make it easy for anyone to add new translations, they just have to add a new entry in  m_languagesAndCodes object of the singleton class Translator and any other thing will update. (exemple : m_languagesAndCodes["deutsh"] = "de"). And they don't have to understand the big MainWindow class.

 I know it's not everyone that likes Singleton design pattern but I thought the Translator class was a good candidate for that and translation shouldn't be tied to a single window like MainWindow it should be tied to the application itself. 

More generally, I think that giving all responsabilities to MainWindow.cpp and having all UI in the same class MainWindow.ui isn't the best design and it might make it harder for Johnny to expand. Maybe putting some refactoring this summer in the timeline might be a good idea.

The mainwindow.ui changed a lot probably because Qt Designer 5.4 write the code differently than 4.X. Don't worry, I didn't mess anything with this file, I just added a layout, an empty comboBox and a label in the designer. 

PS: If you want to test, you have to uncheck the apply settings as soon as they are changed checkbox, I didn't do it yet. It's not hard but I really need to go study.I don't guarantee that there are no other bugs as this is still a work in progress. I just wanted to give you an update as soon to show you my progress.

File: main.cpp [hunk 1]
```diff
@@ -3,10 +3,11 @@
...
  */
 
-#include <QtCore/QCoreApplication>
-#include <QtWidgets/QApplication>
+#include <QCoreApplication>
+#include <QApplication>
+#include <QDir>
 #include "mainwindow.h"
-
+#include "translator.h"
```
Comment (shinnok) [inline main.cpp]: Let's move private file includes to the top and the Qt ones bellow. As per the coding style rules we should arrange include directives in descending order of specificity. http://doc.qt.digia.com/qtcreator-extending/coding-style.html#including-headers

File: main.cpp [hunk 2]
```diff
@@ -19,6 +20,24 @@ int main(int argc, char *argv[])
 
     QApplication app(argc, argv);
 
+    QSettings settings(QDir(QDir::home().filePath(".john")).filePath("johnny.conf"),
```
Comment (shinnok) [inline main.cpp]: Let's pass this settings var to MainWindow instead of initializing another one for that.

**Comment by AlekseyCherepanov** [comment]
Shinnok - yes, I commented out parts of the xml manually in my text editor. The change is not from some qt's tool. I think they don't keep unneeded code.

**Comment by math07** [comment]
Your suggestions are great, I'll correct them on monday.

**Comment by math07** [comment]
Done :)

**Comment by shinnok** [comment]
@math07 There's a couple more refinements to do before I can merge this. See latest comments. Thanks.

**Comment by math07** [comment]
Changes done ! Since QSettings(const& ref) is a private constructor, I had to keep QSettings& as class member instead of QSettings if we didn't want QSettings*. I'm not 100% sure if that's what you meant :)

**Pull request merged**